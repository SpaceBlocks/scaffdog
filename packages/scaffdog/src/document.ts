import path from 'path';
import type {
  ExtractOptions,
  Template,
  VariableSourceMap,
} from '@scaffdog/core';
import { extract } from '@scaffdog/core';
import { ScaffdogError } from '@scaffdog/error';
import frontmatter from 'front-matter';
import globby from 'globby';
import * as z from 'zod';
import { readFile } from './utils/fs';

const MARKDOWN_EXTNAME = new Set([
  '.markdown',
  '.md',
  '.mdown',
  '.mdtext',
  '.mdtxt',
  '.mdwn',
  '.mkd',
  '.mkdn',
]);

const questionIfSchema = z.union([z.boolean(), z.string()]);

const questionSchema = z.union([
  // input syntax sugar
  z.string(),
  // checkbox
  z
    .object({
      message: z.string(),
      choices: z.array(z.string()),
      multiple: z.boolean().optional(),
      initial: z.array(z.string()).optional(),
      if: questionIfSchema.optional(),
    })
    .strict(),
  // list
  z
    .object({
      message: z.string(),
      choices: z.array(z.string()),
      multiple: z.boolean().optional(),
      initial: z.string().optional(),
      if: questionIfSchema.optional(),
    })
    .strict(),
  // confirm
  z
    .object({
      confirm: z.string(),
      initial: z.boolean().optional(),
      if: questionIfSchema.optional(),
    })
    .strict(),
  // input
  z
    .object({
      message: z.string(),
      initial: z.string().optional(),
      if: questionIfSchema.optional(),
    })
    .strict(),
]);

const attrSchema = z.object({
  name: z.string(),
  root: z.string(),
  output: z.union([z.string(), z.array(z.string())]),
  ignore: z.array(z.string()).optional(),
  questions: z.record(questionSchema).optional(),
});

export type Question = z.infer<typeof questionSchema>;
export type DocumentAttributes = z.infer<typeof attrSchema>;

export type Document = DocumentAttributes & {
  path: string;
  templates: Template[];
  variables: VariableSourceMap;
};

export const parseDocument = (
  path: string,
  input: string,
  options: ExtractOptions,
): Document => {
  const { attributes, body } = frontmatter<DocumentAttributes>(input);
  const attrs = attrSchema.safeParse(attributes);
  if (!attrs.success) {
    const [issue] = attrs.error.issues;
    const paths = issue.path.join('.');
    const msg = `Document Parsing Error: in '${paths}': ${issue.message}`;
    throw new ScaffdogError(msg, {
      source: input,
    });
  }

  const { variables, templates } = extract(body, options);

  return {
    ...attrs.data,
    path,
    variables,
    templates,
  };
};

export const resolveDocuments = async (
  dirname: string,
  patterns: string[],
  options: ExtractOptions,
): Promise<Document[]> => {
  const paths = await globby(patterns, {
    cwd: dirname,
    onlyFiles: true,
    unique: true,
    dot: true,
    absolute: true,
  });

  return await Promise.all(
    paths
      .filter((filepath) => MARKDOWN_EXTNAME.has(path.extname(filepath)))
      .map(async (filepath) => {
        const content = await readFile(filepath, 'utf8');
        return parseDocument(filepath, content, options);
      }),
  );
};
