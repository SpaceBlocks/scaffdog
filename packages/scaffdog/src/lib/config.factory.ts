import merge from 'deepmerge';
import { isPlainObject } from 'is-plain-object';
import type { ResolvedConfig } from '@scaffdog/types';
import type { PartialDeep } from 'type-fest';

export const createResolvedConfig = (
  props: PartialDeep<ResolvedConfig> = {},
): ResolvedConfig =>
  merge(
    {
      files: [],
      variables: new Map(),
      helpers: new Map(),
    },
    props as any,
    {
      isMergeableObject: isPlainObject,
    },
  );
