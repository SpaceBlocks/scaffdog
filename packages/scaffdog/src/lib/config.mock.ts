// NOTE: This is auto-generated file.
import { vi } from 'vitest';
import type { ConfigLibrary } from './config';
export const createConfigLibraryMock = (
  mocks: { [P in keyof ConfigLibrary]?: any } = {},
): ConfigLibrary => ({
  load: vi.fn(),
  ...mocks,
});
