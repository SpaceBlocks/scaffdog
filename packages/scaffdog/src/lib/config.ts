import type { LoadConfigResult } from '@scaffdog/config';
import { loadConfig } from '@scaffdog/config';
import type { Consola } from 'consola';
import type { ErrorLibrary } from './error';

export type ConfigLibrary = {
  load: (project: string) => LoadConfigResult | null;
};

export const createConfigLibrary = (
  logger: Consola,
  error: ErrorLibrary,
): ConfigLibrary => ({
  load: (project) => {
    try {
      const result = loadConfig(project);
      logger.debug('loaded config path: %s', result.filepath);
      logger.debug('loaded config: %O', result.config);
      return result;
    } catch (e) {
      error.handle(e, 'Load Config Error');
      return null;
    }
  },
});
