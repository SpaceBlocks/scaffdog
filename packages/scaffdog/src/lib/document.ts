import path from 'path';
import type {
  ExtractOptions,
  Template,
  VariableSourceMap,
} from '@scaffdog/core';
import { extract } from '@scaffdog/core';
import { ScaffdogError } from '@scaffdog/error';
import frontmatter from 'front-matter';
import * as z from 'zod';
import type { FsLibrary } from './fs';
import { questionRecordSchema } from './question';

const MARKDOWN_EXTNAME = new Set([
  '.markdown',
  '.md',
  '.mdown',
  '.mdtext',
  '.mdtxt',
  '.mdwn',
  '.mkd',
  '.mkdn',
]);

export type DocumentAttributes = z.infer<typeof attrSchema>;
const attrSchema = z.object({
  name: z.string(),
  root: z.string(),
  output: z.union([z.string(), z.array(z.string())]),
  ignore: z.array(z.string()).optional(),
  questions: questionRecordSchema.optional(),
});

export type Document = DocumentAttributes & {
  path: string;
  templates: Template[];
  variables: VariableSourceMap;
};

/**
 * @internal
 */
export const parse = (
  path: string,
  input: string,
  options: ExtractOptions,
): Document => {
  const { attributes, body } = frontmatter<DocumentAttributes>(input);
  const attrs = attrSchema.safeParse(attributes);
  if (!attrs.success) {
    const [issue] = attrs.error.issues;
    const paths = issue.path.join('.');
    const msg = `Document Parsing Error: in '${paths}': ${issue.message}`;
    throw new ScaffdogError(msg, {
      source: input,
    });
  }

  const { variables, templates } = extract(body, options);

  return {
    ...attrs.data,
    path,
    variables,
    templates,
  };
};

/**
 * @public
 */
export type DocumentLibrary = {
  resolve: (
    dirname: string,
    patterns: string[],
    options: ExtractOptions,
  ) => Promise<Document[]>;
};

export const createDocumentLibrary = (fs: FsLibrary): DocumentLibrary => ({
  resolve: async (dirname, patterns, options) => {
    const paths = await fs.glob(patterns, {
      cwd: dirname,
      onlyFiles: true,
      unique: true,
      dot: true,
      absolute: true,
    });

    return await Promise.all(
      paths
        .filter((filepath) => MARKDOWN_EXTNAME.has(path.extname(filepath)))
        .map(async (filepath) => {
          const content = await fs.readFile(filepath);
          return parse(filepath, content, options);
        }),
    );
  },
});

createDocumentLibrary.inject = ['fs'] as const;
