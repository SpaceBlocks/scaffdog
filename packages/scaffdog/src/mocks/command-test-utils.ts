import path from 'path';
import type {
  AnyCommandModule,
  CommandArgs,
  CommandModule,
  CommandOption,
  GlobalFlags,
  GlobalFlagsWith,
  InferredCommandModuleArgs,
  InferredCommandModuleFlags,
} from '../command';
import type { CommandContainer } from '../command-container';
import { createCommandContainer } from '../command-container';
import type { Library } from '../lib';
import { createLibraryMock } from './lib';
import { createLogger } from './logger';

export const cwd = path.resolve(__dirname, '../../');

export type CommandRunnerParams<T extends AnyCommandModule> = {
  args?: Partial<CommandArgs<InferredCommandModuleArgs<T>>>;
  flags?: Partial<GlobalFlagsWith<CommandArgs<InferredCommandModuleFlags<T>>>>;
  lib?: Library;
  container?: CommandContainer;
};

export type CommandRunner<T extends AnyCommandModule> = (
  params: CommandRunnerParams<T>,
) => Promise<{
  code: number;
  stdout: string;
  stderr: string;
}>;

export const createCommandRunner =
  <A extends CommandOption, F extends CommandOption>(
    cmd: CommandModule<A, F>,
    defaults: {
      args: CommandArgs<A>;
      flags: CommandArgs<F>;
    },
  ): CommandRunner<CommandModule<A, F>> =>
  async ({ args, flags, lib, container }) => {
    const { logger, getStdout, getStderr } = createLogger();

    const globalFlags: GlobalFlags = {
      project: '.scaffdog',
      help: false,
      version: false,
      verbose: true,
    };

    const code = await cmd.run({
      cwd,
      pkg: {
        name: 'scaffdog',
        version: '1.0.0',
        description: 'scaffdog is Markdown driven scaffolding tool.',
      },
      size: {
        rows: 120,
        columns: 60,
      },
      logger,
      container: container ?? createCommandContainer([]),
      lib: lib ?? createLibraryMock({}),
      args: {
        ...defaults.args,
        ...(args ?? {}),
      },
      flags: {
        ...defaults.flags,
        ...globalFlags,
        ...(flags ?? {}),
      } as never,
    });

    return {
      code,
      stdout: getStdout(),
      stderr: getStderr(),
    };
  };
