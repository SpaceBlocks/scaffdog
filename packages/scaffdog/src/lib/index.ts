import type { ConsolaInstance } from 'consola';
import { createConfigLibrary } from './config.js';
import { createDocumentLibrary } from './document.js';
import { createErrorLibrary } from './error.js';
import { createFsLibrary } from './fs.js';
import { createPromptLibrary } from './prompt.js';
import { createQuestionLibrary } from './question.js';

export const createLibrary = (logger: ConsolaInstance) => {
  const fs = createFsLibrary();
  const error = createErrorLibrary(logger);
  const prompt = createPromptLibrary();
  const config = createConfigLibrary(logger, error);
  const question = createQuestionLibrary(prompt);
  const document = createDocumentLibrary(fs, question);

  return {
    fs,
    error,
    prompt,
    config,
    question,
    document,
  };
};

export type Library = ReturnType<typeof createLibrary>;
