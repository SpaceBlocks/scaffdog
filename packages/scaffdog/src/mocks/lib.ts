import type { Library } from '../lib';
import { createConfigLibraryMock } from '../lib/config.mock';
import { createDocumentLibraryMock } from '../lib/document.mock';
import { createErrorLibraryMock } from '../lib/error.mock';
import { createFsLibraryMock } from '../lib/fs.mock';
import { createPromptLibraryMock } from '../lib/prompt.mock';
import { createQuestionLibraryMock } from '../lib/question.mock';

export const createLibraryMock = (mocks: Partial<Library> = {}): Library => {
  return {
    fs: createFsLibraryMock(),
    error: createErrorLibraryMock(),
    prompt: createPromptLibraryMock(),
    config: createConfigLibraryMock(),
    question: createQuestionLibraryMock(),
    document: createDocumentLibraryMock(),
    ...mocks,
  };
};
