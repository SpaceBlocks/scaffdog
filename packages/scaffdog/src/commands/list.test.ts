import path from 'path';
import { parse } from '@scaffdog/engine';
import { describe, expect, test, vi } from 'vitest';
import { createResolvedConfig } from '../lib/config.factory';
import { createConfigLibraryMock } from '../lib/config.mock';
import { createDocument } from '../lib/document.factory';
import { createDocumentLibraryMock } from '../lib/document.mock';
import { createCommandRunner, cwd } from '../mocks/command-test-utils';
import { createLibraryMock } from '../mocks/lib';
import cmd from './list';

const config = createConfigLibraryMock({
  load: () =>
    createResolvedConfig({
      files: ['./*'],
      tags: ['{{', '}}'],
    }),
});

const run = createCommandRunner(cmd, {
  args: {},
  flags: {},
});

test('success', async () => {
  const documents = [
    createDocument({
      name: 'doc1',
      path: path.join(cwd, '.scaffdog/doc1.md'),
      templates: [
        {
          filename: parse('1'),
          content: parse('1'),
        },
      ],
    }),
    createDocument({
      name: 'doc2',
      path: path.join(cwd, '.scaffdog/doc2.md'),
      templates: [
        {
          filename: parse('1'),
          content: parse('1'),
        },
        {
          filename: parse('2'),
          content: parse('2'),
        },
      ],
      questions: {
        key1: 'msg',
      },
    }),
    createDocument({
      name: 'doc3',
      path: path.join(cwd, '.scaffdog/doc3.md'),
      templates: [
        {
          filename: parse('1'),
          content: parse('1'),
        },
      ],
      questions: {
        key1: 'msg',
        key2: 'msg',
      },
    }),
  ];

  const lib = createLibraryMock({
    config,
    document: createDocumentLibraryMock({
      resolve: vi.fn().mockResolvedValueOnce(documents),
    }),
  });

  const { code, stdout, stderr } = await run({
    lib,
  });

  expect(code).toBe(0);
  expect(stderr).toBe('');
  expect(stdout).toMatchSnapshot();

  expect(lib.document.resolve).toBeCalledWith(
    path.resolve(cwd, '.scaffdog'),
    ['./*'],
    {
      tags: ['{{', '}}'],
    },
  );
});

describe('failure', () => {
  test('config not found', async () => {
    const lib = createLibraryMock({
      config: createConfigLibraryMock({
        load: vi.fn().mockReturnValueOnce(null),
      }),
    });

    const { code, stdout, stderr } = await run({
      lib,
    });

    expect(code).toBe(1);
    expect(stderr).toBe('');
    expect(stdout).toBe('');
  });

  test('document not found', async () => {
    const lib = createLibraryMock({
      config,
      document: createDocumentLibraryMock({
        resolve: vi.fn().mockResolvedValueOnce([]),
      }),
    });

    const { code, stdout, stderr } = await run({
      lib,
    });

    expect(code).toBe(1);
    expect(stderr).toMatchSnapshot();
    expect(stdout).toBe('');
  });
});
