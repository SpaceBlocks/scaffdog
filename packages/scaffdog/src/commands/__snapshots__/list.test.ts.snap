// Vitest Snapshot v1

exports[`failure > document not found 1`] = `
"
 WARN  Document file not found.

"
`;

exports[`success 1`] = `
"- doc1 (.scaffdog/doc1.md, 1 template, 0 questions)
- doc2 (.scaffdog/doc2.md, 2 templates, 1 question)
- doc3 (.scaffdog/doc3.md, 1 template, 2 questions)

3 files found.
"
`;
